File: packages/sdk/typescript/src/integration-adapter.ts
import type { RelayFileClient } from "./client.js";
import type { ConnectionProvider } from "./connection.js";
import type { FileSemantics } from "./types.js";
import type { WriteEvent } from "@relayfile/core";

export interface AdapterWebhookMetadata {
  deliveryId?: string;
  delivery_id?: string;
  timestamp?: string;
  [key: string]: unknown;
}

export interface AdapterWebhook {
  provider: string;
  connectionId?: string;
  eventType: string;
  objectType: string;
  objectId: string;
  payload: Record<string, unknown>;
  metadata?: AdapterWebhookMetadata;
  raw?: unknown;
}

export interface IngestError {
  path: string;
  error: string;
}

export interface IngestResult {
  filesWritten: number;
  filesUpdated: number;
  filesDeleted: number;
  paths: string[];
  errors: IngestError[];
}

export interface SyncOptions {
  cursor?: string;
  limit?: number;
  signal?: AbortSignal;
  [key: string]: unknown;
}

export interface SyncResult {
  filesWritten: number;
  filesUpdated: number;
  filesDeleted: number;
  paths?: string[];
  cursor?: string;
  nextCursor?: string | null;
  syncedObjectTypes?: string[];
  errors: Array<{
    path?: string;
    objectType?: string;
    error: string;
  }>;
}

export interface RelayBindingView {
  text: string;
  author?: string;
  skip?: boolean;
}

export interface RelayBinding {
  present?(event: WriteEvent): RelayBindingView | null;
  replyPathFor?(sourcePath: string): string | null;
}

const DEFAULT_PRESENT_TEXT_LIMIT = 500;

export abstract class IntegrationAdapter {
  protected readonly client: RelayFileClient;
  protected readonly provider: ConnectionProvider;

  abstract readonly name: string;
  abstract readonly version: string;

  constructor(client: RelayFileClient, provider: ConnectionProvider) {
    this.client = client;
    this.provider = provider;
  }

  abstract ingestWebhook(workspaceId: string, event: AdapterWebhook): Promise<IngestResult>;
  abstract computePath(objectType: string, objectId: string): string;
  abstract computeSemantics(
    objectType: string,
    objectId: string,
    payload: Record<string, unknown>,
  ): FileSemantics;
  supportedEvents?(): string[];
  writeBack?(workspaceId: string, path: string, content: string): Promise<unknown>;
  relayBinding?: RelayBinding;
  sync?(workspaceId: string, options?: SyncOptions): Promise<SyncResult>;
}

export function genericPresent(event: WriteEvent, adapter: IntegrationAdapter): RelayBindingView {
  const provider = providerFromPath(event.path) || adapter.name;
  const { objectType, objectId } = objectIdentityFromPath(event.path);
  const payload = isRecord(event.value) ? event.value : {};
  let semantics: FileSemantics = {};
  try {
    semantics = adapter.computeSemantics(objectType, objectId, payload) ?? {};
  } catch {
    semantics = {};
  }

  const summary = summarizeForRelayBinding(event, semantics, payload);
  return {
    text: truncateText(summary, DEFAULT_PRESENT_TEXT_LIMIT),
    author: event.actor?.id || provider
  };
}

export function genericReplyPathFor(sourcePath: string): string | null {
  const normalized = sourcePath.trim().replace(/\/+$/, "");
  if (!normalized || !normalized.startsWith("/")) {
    return null;
  }
  return `${normalized}/replies/draft.json`;
}

function summarizeForRelayBinding(
  event: WriteEvent,
  semantics: FileSemantics,
  payload: Record<string, unknown>
): string {
  const semanticLines = [
    firstNonEmpty("status", semantics.properties?.["provider.status"], semantics.properties?.status),
    firstNonEmpty("updated", semantics.properties?.["provider.updated_at"], semantics.properties?.updatedAt),
    firstNonEmpty("title", semantics.properties?.title, semantics.properties?.name, semantics.properties?.subject)
  ].filter((line): line is string => line !== undefined);

  const relationCount = semantics.relations?.length ?? 0;
  const commentCount = semantics.comments?.length ?? 0;
  const permissionCount = semantics.permissions?.length ?? 0;
  const counts = [
    relationCount ? `${relationCount} relation${relationCount === 1 ? "" : "s"}` : undefined,
    commentCount ? `${commentCount} comment${commentCount === 1 ? "" : "s"}` : undefined,
    permissionCount ? `${permissionCount} permission${permissionCount === 1 ? "" : "s"}` : undefined
  ].filter((entry): entry is string => entry !== undefined);

  const payloadPreview = Object.keys(payload).length > 0
    ? `payload: ${JSON.stringify(payload)}`
    : undefined;

  return [
    `${event.operation} ${event.path}`,
    counts.length > 0 ? counts.join(", ") : undefined,
    ...semanticLines,
    payloadPreview
  ].filter((line): line is string => line !== undefined && line.trim() !== "").join("\n");
}

function firstNonEmpty(label: string, ...values: Array<string | undefined>): string | undefined {
  const value = values.find((candidate) => candidate && candidate.trim() !== "");
  return value ? `${label}: ${value}` : undefined;
}

function truncateText(value: string, limit: number): string {
  if (value.length <= limit) {
    return value;
  }
  return value.slice(0, limit);
}

function providerFromPath(path: string): string {
  return normalizePath(path)[0] ?? "";
}

function objectIdentityFromPath(path: string): { objectType: string; objectId: string } {
  const segments = normalizePath(path);
  const filename = segments[segments.length - 1] ?? "";
  return {
    objectType: segments[1] ?? "unknown",
    objectId: filename.replace(/\.[^.]+$/, "")
  };
}

function normalizePath(path: string): string[] {
  return path.split("/").filter(Boolean);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

File: packages/sdk/python/src/relayfile/integration_adapter.py
from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, Protocol, runtime_checkable

from .client import RelayFileClient
from .connection import ConnectionProvider
from .types import FileSemantics, WriteEvent, WriteEventActor


@dataclass
class AdapterWebhook:
    provider: str
    event_type: str
    object_type: str
    object_id: str
    payload: dict[str, Any]
    connection_id: str | None = None
    metadata: dict[str, Any] | None = None
    raw: Any = None


@dataclass
class IngestError:
    path: str
    error: str


@dataclass
class IngestResult:
    files_written: int
    files_updated: int
    files_deleted: int
    paths: list[str] = field(default_factory=list)
    errors: list[IngestError] = field(default_factory=list)


@dataclass
class SyncOptions:
    cursor: str | None = None
    limit: int | None = None
    extra: dict[str, Any] = field(default_factory=dict)


@dataclass
class SyncError:
    error: str
    path: str | None = None
    object_type: str | None = None


@dataclass
class SyncResult:
    files_written: int
    files_updated: int
    files_deleted: int
    paths: list[str] | None = None
    cursor: str | None = None
    next_cursor: str | None = None
    synced_object_types: list[str] | None = None
    errors: list[SyncError] = field(default_factory=list)


@dataclass
class RelayBindingView:
    text: str
    author: str | None = None
    skip: bool = False


@runtime_checkable
class RelayBinding(Protocol):
    def present(self, event: WriteEvent) -> RelayBindingView | None: ...
    def reply_path_for(self, source_path: str) -> str | None: ...


_DEFAULT_PRESENT_TEXT_LIMIT = 500


def generic_present(event: WriteEvent, adapter: IntegrationAdapter) -> RelayBindingView:
    provider = _provider_from_path(event.path) or adapter.name
    payload: dict[str, Any] = event.value if isinstance(event.value, dict) else {}
    try:
        object_type, object_id = _object_identity_from_path(event.path)
        semantics = adapter.compute_semantics(object_type, object_id, payload)
    except Exception:
        semantics = FileSemantics()

    lines: list[str] = [f"{event.operation} {event.path}"]
    text = "\n".join(line for line in lines if line.strip())
    actor = event.actor
    author: str | None = None
    if isinstance(actor, WriteEventActor):
        author = actor.id
    elif isinstance(actor, dict):
        author = str(actor.get("id", "")) or None
    return RelayBindingView(
        text=text[:_DEFAULT_PRESENT_TEXT_LIMIT],
        author=author or provider,
    )


def generic_reply_path_for(source_path: str) -> str | None:
    normalized = source_path.strip().rstrip("/")
    if not normalized or not normalized.startswith("/"):
        return None
    return f"{normalized}/replies/draft.json"


def _provider_from_path(path: str) -> str:
    segments = [s for s in path.split("/") if s]
    return segments[0] if segments else ""


def _object_identity_from_path(path: str) -> tuple[str, str]:
    segments = [s for s in path.split("/") if s]
    object_type = segments[1] if len(segments) > 1 else "unknown"
    filename = segments[-1] if segments else ""
    object_id = filename.rsplit(".", 1)[0] if "." in filename else filename
    return object_type, object_id


class IntegrationAdapter(ABC):
    name: str
    version: str
    relay_binding: RelayBinding | None = None

    def __init__(self, client: RelayFileClient, provider: ConnectionProvider) -> None:
        self.client = client
        self.provider = provider

    @abstractmethod
    def ingest_webhook(self, workspace_id: str, event: AdapterWebhook) -> IngestResult:
        ...

    @abstractmethod
    def compute_path(self, object_type: str, object_id: str) -> str:
        ...

    @abstractmethod
    def compute_semantics(
        self,
        object_type: str,
        object_id: str,
        payload: dict[str, Any],
    ) -> FileSemantics:
        ...

    def supported_events(self) -> list[str] | None:
        return None

    def write_back(self, workspace_id: str, path: str, content: str) -> Any:
        raise NotImplementedError

    def sync(self, workspace_id: str, options: SyncOptions | None = None) -> SyncResult:
        raise NotImplementedError
